schema-lending.graphql

```diff
@@ -141,6 +155,101 @@ type InterestRate @entity {
   type: InterestRateType!
 }
 
+# There is a wide range of formats of lending protocols
+# And there are many unique mappings when it comes to input / output tokens
+# This generic type will help define the unique relationships between tokens in a market
+type TokenBalances @entity {
+  " { Market Address }-{ Optional: Snapshot ID } "
+  id: ID!
+
+  " Tokens that need to be deposited in this market to take a position in protocol (should be alphabetized) "
+  inputTokens: [Token!]!
+
+  " Amount of input tokens in the market (same order as inputTokens) "
+  inputTokenBalances: [BigInt!]!
+
+  " Prices in USD of the input tokens (same order as inputTokens) "
+  inputTokenPricesUSD: [BigDecimal!]!
+
+  " Tokens that are minted to track ownership of position in protocol (e.g. aToken, cToken). Leave as null if doesn't exist (should be alphabetized) "
+  outputTokens: [Token!]
+
+  " Total supply of output tokens (same order as outputTokens) "
+  outputTokenBalances: [BigInt!]
+
+  " Prices in USD of the output tokens (same order as outputTokens) "
+  outputTokenPricesUSD: [BigDecimal!]
+
+  " Amount of input token per full share of output token. Only applies when the output token exists (note this is a ratio and not a percentage value, i.e. 1.05 instead of 105%) "
+  exchangeRates: [BigDecimal!]
+}
+
+enum RevenueSource {
+  " Revenue that is generated from borrower's interest "
+  BORROW_INTEREST
+
+  " Revenue that is generated from liquidation fees "
+  LIQUIDATION_FEE
+
+  " Revenue that is generated from fees from flashloans "
+  FLASHLOAN_FEE
+}
+
+# This entity offers a more nuanced view of revenue types
+# Use this to provide the sources of revenue and amounts of each
+type RevenueDetails @entity {
+  " { Market/Protocol ID } "
+  id: ID!
```

Comment:
consider making this market-revenue type so that we don't have to keep the arrays in a specific order, then adding a market field here to derive this into each market.

Should this be at the protocol level?

Reply:
done